cvat-cli/src/cvat_cli/_internal/agent.py

```diff
@@ -0,0 +1,351 @@
+                body={"agent_id": self._agent_id, "annotations": result},
+            )
+            self._client.logger.info("AR %r completed", ar_id)
+        except Exception as ex:
+            self._client.logger.error("Failed to process AR %r", ar_id, exc_info=True)
+
+            # Arbitrary exceptions may contain details of the client's system or code, which
+            # shouldn't be exposed to the server (and to users of the function).
+            # Therefore, we only produce a limited amount of detail, and only in known failure cases.
+            error_message = "Unknown error"
+
+            if isinstance(ex, ApiException):
+                if ex.status:
+                    error_message = f"Received HTTP status {ex.status}"
+                else:
+                    error_message = "Failed an API call"
+            elif isinstance(ex, urllib3.exceptions.RequestError):
+                if isinstance(ex, urllib3.exceptions.MaxRetryError):
+                    ex_type = type(ex.reason)
+                else:
+                    ex_type = type(ex)
+
+                error_message = f"Failed to make an HTTP request to {ex.url} ({ex_type.__name__})"
+            elif isinstance(ex, urllib3.exceptions.HTTPError):
+                error_message = "Failed to make an HTTP request"
+            elif isinstance(ex, cvataa.BadFunctionError):
+                error_message = "Underlying function returned incorrect result: " + str(ex)
+            elif isinstance(ex, concurrent.futures.BrokenExecutor):
+                error_message = "Worker process crashed"
+
+            try:
+                self._client.api_client.call_api(
+                    "/api/functions/queues/{queue_id}/requests/{request_id}/fail",
+                    "POST",
+                    path_params={
+                        "queue_id": f"function:{self._function_id}",
+                        "request_id": ar_id,
+                    },
+                    body={"agent_id": self._agent_id, "exc_info": error_message},
+                )
+            except Exception:
+                self._client.logger.error("Couldn't fail AR %r", ar_id, exc_info=True)
+            else:
+                self._client.logger.info("AR %r failed", ar_id)
+
+    def _poll_for_ar(self) -> Optional[dict]:
+        while True:
+            self._client.logger.info("Trying to acquire an annotation request...")
+            try:
+                _, response = self._client.api_client.call_api(
+                    "/api/functions/queues/{queue_id}/requests/acquire",
+                    "POST",
+                    path_params={"queue_id": f"function:{self._function_id}"},
+                    body={"agent_id": self._agent_id, "request_category": "batch"},
+                )
+                break
+            except (urllib3.exceptions.HTTPError, ApiException) as ex:
+                if isinstance(ex, ApiException) and ex.status and 400 <= ex.status < 500:
```

Comment:
Consider skipping as well:
- 408 Request Timeout - for server-side problems (I'd add a refreshable limited retry counter for this)
- 429 Too Many Requests - for server-side throttling

This should make agents less faulty because of various server performance problems.

Reply:
Ok, then instead of 408 it should simply be the urllib3 timeout exception, I guess. Please check how it's handled.

Reply:
urllib3's `TimeoutError` inherits from `HTTPError`, so it would be caught here. Since it's not an `ApiException`, it would not be re-raised, so the code will try again.